### PR TITLE
fix(rmp): RateScheduler unscheduling not working

### LINKED ISSUE #3302: [BUG] CoherentGTDraw gets very high after long flight
**Mod Version**

latest master commit

**Describe the bug**

**To Reproduce**
1. Perform a long (>2hr) flight

**Expected behavior**

* CoherentGTDraw stays consistent throughout the flight

**Actual behavior**

* CoherentGTDraw slowly gets higher

Discord username (if different from GitHub): someperson#4953

### PR DESCRIPTION
**Pull request by theomessin**
Fixes #3302

## Summary of Changes
Changed RateScheduler to identify callbacks by an id as opposed to the callback itself. This means that unscheduling now works.

Discord username (if different from GitHub): Theodore Messinezis#6029

## Testing instructions
See #3302

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

File: src/instruments/src/RMP/Framework/RateScheduler.mjs [hunk 1]
```diff
@@ -4,33 +4,38 @@
...
+        this.rates.set(rate, {
             interval: setInterval(() => this.tick(rate), rate),
-            callbacks: [callback],
+            identifiers: [identifier],
         });
     }
 
     /**
      * Remove a given callback (from all rates).
      * @param callback The callback to unschedule.
```
Comment (theomessin) [inline src/instruments/src/RMP/Framework/RateScheduler.mjs:32]: Need to update jsdoc block.

File: src/instruments/src/RMP/Framework/RateScheduler.mjs [hunk 2]
```diff
@@ -4,33 +4,38 @@
...
     constructor() {
-        this.map = new Map();
+        this.rates = new Map();
+        this.callbacks = new Map();
     }
 
     /**
      * Schedule a new callback to be called at a set rate.
      * @param callback The callback to be scheduled
      * @param rate The rate that this callback should be called at.
```
Comment (theomessin) [inline src/instruments/src/RMP/Framework/RateScheduler.mjs:14]: Need to update jsdoc block.

File: src/instruments/src/RMP/Framework/RateScheduler.mjs [hunk 3]
```diff
@@ -40,8 +45,8 @@ export class RateScheduler {
      * Clears intervals that have no callbacks.
```
Comment (theomessin) [inline src/instruments/src/RMP/Framework/RateScheduler.mjs:45]: Need to update jsdoc block.

**Comment by GlitchedYT** [comment]
Category: Unofficial QA Tester
Name: Glitched#0740
Date of testing: 2/5/2021
Version of the sim: 1.12.13.0
PR Tested: #3358
Tier of testing: Tier 2
Changes to observe: Changed RateScheduler to identify callbacks by an id as opposed to the callback itself. This means that unscheduling now works.
(Insert the Summary of changes, in order to remind what is to test)

Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:
1. Spawned C&D at KDEN gate B38
2. Loaded flight as i would normally do through simbrief (KDEN-KORD)
3. Took off, climbed to 370 and started to mess around with both RMP's as much as I could.
4. Observed any frame drops/stutters/increase in frame times when messing with both RMP's
5. Performed arrival and approach perfectly and finished the flight.

Flight notes (When done a full flight):
Flight went very well, only frame drops were due to FSDreamTeam KORD

Media: Insert screenshots, recordings, or anything relevant to the review here.
KDEN (start of flight with all displays turned on)
https://i.imgur.com/OgmDJdg.jpg
KORD (end of flight, about 2 hours later)
https://i.imgur.com/w98mRPr.jpg

Issues (If Any): N/A

Overall Rating: Good, no more (noticable) performance issues with the RMP. All flight i was messing with it as much as i could but nothing crazy happened.
Conclusions of the testing: Custom RMP is now usable from my end after using this fix, thanks you to @XSJoJo and @Theodore Messinezis for getting this fixed so quickly.

**Comment by sidnov** [comment]
Category: QA Tester
Name: sidnov#8337
Date of testing: 6.2.2021
Version of the sim: 1.12.13.0
PR Tested: #3358 
Tier of testing: 1

### Changes to observe: RateScheduler fixes FPS/CoherentGTDraw

### Testing technique: Tryout

### Steps you've made to test the Pull Request:
1. Spawned C/D at VABB, gate 30
2. Completed preflight as normal. Used external power, APU.
3. RMPs were ON and tuned to 122.95 and 121.5. Connected to POSCON network.
4. Tested all radios mid-flight. Worked as expected.
5. FPS and CoherentGTDraw changes were observed.
6. 2.5 hours into the flight, FPS was observed, screenshot attached.
7. (To be updated)

### Media: image

### Issues (If Any): (To be updated)

### Conclusions of the testing: (To be updated)

**Comment by InhumanReactions** [comment]
Category: QA Tester
Name: Bowling Ball#7947
Date of testing: Current date 06/02/2021
Version of the sim: 1.12.13.0
PR Tested: #3358
Tier of testing: 2
Changes to observe: Change in frame rate after around 1h 40m of flight.

Testing technique: I completed a full flight.
Steps you've made to test the Pull Request:
1. Spawned in C&D at EGLL Stand 519 for a flight to LXGB.
2. Set up the flight and had to reroute via request from controller.
3. Took off and played around with the RMP panels (with no noticeable frame drop).
4. Monitored the FPS throughout the flight.
5. Descended and had to conduct a go-around.
6. Shut down the aircraft and monitored the FPS.

Flight notes: Flight time was 2h 40m. I had general spikes in FPS but these occurred throughout the whole flight and therefore not an issue of the PR itself.

Media: 
FlightSimulator_yLDbS9pEqX 
FlightSimulator_1KfMMH4Swz



Overall Rating: Good.
Conclusions of the testing: This PR worked as expected and the FPS remained constant throughout alongside the latency of CoherantGTDraw. Well done on getting the issue fixed!

**Comment by FlightSimmer7700** [comment]
Category: QA Tester
Name: FlightSimmer7700#3468
Date of testing: 6/2/21
Version of the sim: 1.12.13.0
PR Tested: #3358 
Tier of testing: 1/2/3
Changes to observe:
coherent gt draw

Testing technique: did a full flight
Steps you've made to test the Pull Request:
1. did a full flight


Flight notes (When done a full flight):

Media: Insert screenshots, recordings, or anything relevant to the review here.
Screenshot (478)
Screenshot (480)
Screenshot (481)
Screenshot (482)

Issues (If Any): none
Steps to achieve the issue

Overall Rating: Good
Conclusions of the testing: All good, dont have time to write a full report but fps stayed same and coherent draw too for the whole 2hr flight

**Review by ko-zone — COMMENTED** [review]
LGTM

**Review by ko-zone — APPROVED** [review]

**Pull request merged**